iceoryx_examples/ara_com/ara/runtime.hpp

```diff
@@ -0,0 +1,171 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_EXAMPLES_ARA_COM_RUNTIME_HPP
+#define IOX_EXAMPLES_ARA_COM_RUNTIME_HPP
+
+#include "iceoryx_hoofs/cxx/vector.hpp"
+#include "iceoryx_hoofs/internal/cxx/pair.hpp"
+#include "iceoryx_posh/popo/listener.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "iceoryx_posh/runtime/service_discovery.hpp"
+
+#include "types.hpp"
+
+#include <string>
+
+namespace ara
+{
+class Runtime
+{
+  public:
+    static Runtime& GetInstance(const std::string& name) noexcept
+    {
+        iox::runtime::PoshRuntime::initRuntime(iox::RuntimeName_t(iox::cxx::TruncateToCapacity, name));
+        static Runtime runtime;
+        return runtime;
+    }
+
+    static Runtime& GetInstance() noexcept
+    {
+        static Runtime runtime;
+        return runtime;
+    }
+
+    ara::com::ServiceHandleContainer<com::ProxyHandleType>
+    FindService(ara::com::ServiceIdentifier& serviceIdentifier,
+                ara::com::InstanceIdentifier& instanceIdentifier) noexcept
+    {
+        ara::com::ServiceHandleContainer<com::ProxyHandleType> iceoryxServiceContainer;
+
+        m_discovery.findService(
+            serviceIdentifier,
+            instanceIdentifier,
+            iox::cxx::nullopt,
+            [&](auto& service) {
+                iceoryxServiceContainer.push_back({service.getEventIDString(), service.getInstanceIDString()});
+            },
+            iox::popo::MessagingPattern::PUB_SUB);
+
+        m_discovery.findService(
+            serviceIdentifier,
+            instanceIdentifier,
+            iox::cxx::nullopt,
+            [&](auto& service) {
+                iceoryxServiceContainer.push_back({service.getEventIDString(), service.getInstanceIDString()});
+            },
+            iox::popo::MessagingPattern::REQ_RES);
+
+        // We need to make sure that all three internal services representing 'MinimalSkeleton' are available
+        ara::com::ServiceHandleContainer<com::ProxyHandleType> araServiceContainer;
+        if (verifyThatServiceIsComplete(iceoryxServiceContainer))
+        {
+            araServiceContainer.push_back({serviceIdentifier, instanceIdentifier});
+        }
+
+        return araServiceContainer;
+    }
+
+    ara::com::FindServiceHandle StartFindService(ara::com::FindServiceHandler<com::ProxyHandleType> handler,
+                                                 ara::com::ServiceIdentifier& serviceIdentifier,
+                                                 ara::com::InstanceIdentifier& instanceIdentifier) noexcept
+    {
+        /// @todo #1332 Are duplicate entries allowed?
```

Comment:
Technically it should be no problem to have multiple callbacks for the same service instance pair. Then the handle concept as is would not work though. In general I am not sure whether the handle needs to know about these, as it is mainly/only for (de)registration of callbacks.

Adding an integer would help.